File: app/src/main/java/com/chris/eban/domain/SingleUseCase.java
package com.chris.eban.domain;

import io.reactivex.Single;
import io.reactivex.functions.Function;

/**
 * Single result emitter
 */
public abstract class SingleUseCase<T> implements Interactor<Single<Result<T>>> {


    @Override
    public Single<Result<T>> execute() {
        final Single<T> real = buildSingle();
        if (real == null)
            throw new IllegalArgumentException("Build a UseCase with any params! ");
        return real.map(new Function<T, Result<T>>() {
            @Override
            public Result<T> apply(T t) {
                return new Result<>(t);
            }
        });
    }

    protected abstract Single<T> buildSingle();
}

File: app/src/main/java/com/chris/eban/domain/SingleUseCase.kt
package com.chris.eban.domain

import io.reactivex.Single

/**
 * Single result emitter
 */
abstract class SingleUseCase<T> : Interactor<Single<Result<T>>> {


    override fun execute(): Single<Result<T>> {
        val real = buildSingle()
                ?: throw IllegalArgumentException("Build a UseCase with any params! ")
        return real.map { t -> Result(t) }
    }

    protected abstract fun buildSingle(): Single<T>?
}
